developer/src/kmc-keyboard/src/compiler/touch-layout-compiler.ts

```diff
@@ -0,0 +1,104 @@
+import { TouchLayout, LDMLKeyboard } from "@keymanapp/common-types";
+
+export class TouchLayoutCompiler {
+  public compileToJavascript(source: LDMLKeyboard.LDMLKeyboardXMLSourceFile): TouchLayout.TouchLayoutFile {
+    let result: TouchLayout.TouchLayoutFile = {};
+
+    // start with desktop to mimic vk emit
+    result.desktop = {
+      defaultHint: "none",  // TODO-LDML this should be optional
+      layer: []
+    };
+
+    for(let layers of source.keyboard.layers) {
+      for(let layer of layers.layer) {
+        const resultLayer = this.compileHardwareLayer(source, result, layer);
+        result.desktop.layer.push(resultLayer);
+      }
+    }
+    return result;
+  }
+
+  private compileHardwareLayer(
+    source: LDMLKeyboard.LDMLKeyboardXMLSourceFile,
+    file: TouchLayout.TouchLayoutFile,
+    layer: LDMLKeyboard.LKLayer
+  ) {
+    // TODO-LDML: consider consolidation with keys.ts?
+
+    let fileLayer: TouchLayout.TouchLayoutLayer = {
+      id: this.translateLayerIdToTouchLayoutLayerId(layer.id, layer.modifier),
+      row: []
+    };
+
+    let y = -1;
+
+    for(let row of layer.row) {
+      y++;
+
+      let fileRow: TouchLayout.TouchLayoutRow = {id: y, key: []};
+      fileLayer.row.push(fileRow);
+
+      const keys = row.keys.split(' ');
+      for(let key of keys) {
+        const keydef = source.keyboard.keys?.key?.find(x => x.id == key);
+
+        const fileKey: TouchLayout.TouchLayoutKey = {
+          id: this.translateKeyIdentifierToTouch(keydef.id) as TouchLayout.TouchLayoutKeyId,
+          text: keydef.to,
+          // TODO-LDML: additional properties
+        };
+        fileRow.key.push(fileKey);
+      }
+    }
+
+    return fileLayer;
+  }
+
+  private translateLayerIdToTouchLayoutLayerId(id: string, modifier: string): string {
+    // Touch layout layers have a set of reserved names that correspond to
+    // hardware modifiers. We want to map these identifiers first before falling
+    // back to the layer ids
+
+    // The set of recognized layer identifiers is:
+    //
+    // touch          | LDML
+    // ---------------+-------------
+    // default        | none
+    // shift          | shift
+    // caps           | caps
+    // rightalt       | altR
+    // rightalt-shift | altR shift
+    //
+
+    const map = {
+      none:         'default',
+      shift:        'shift',
+      caps:         'caps',
+      altR:         'rightalt',
+      "altR shift": 'rightalt-shift',
+      "shift altR": 'rightalt-shift', // TODO-LDML is this required?
+    };
+
+    if(Object.hasOwn(map, modifier)) {
+      return (map as any)[modifier];
+    }
+
+    // TODO-LDML: Other layer names will be used unmodified, is this sufficient?
+    return id;
+  }
+
+  private translateKeyIdentifierToTouch(id: string): string {
```

Comment:
shouldn't this be matching on the *text* not assuming the id?

so if we have `<row keys="a z e r t y"/>` the first key would get `K_A` which is fine assuming `<key id='a' to='a'/>` but not fine otherwise.

Reply:
The principle here was to re-use the 0-9, A-Z, a-z identifiers, just for coherence. We could give everything a `T_` prefix, but there is some value in matching these standard identifiers, for example making digit keys visible to the Keyman app.

if `<key id='a' to='foo'/>` is defined, then `K_A` would match to that key, so I think it still works.